src/serde/reader.rs

```diff
@@ -54,15 +58,39 @@ impl<R: Read, E: ByteOrder> Deserializer<R, E> {
         self.read_bytes(size_of::<T>() as u64)
     }
 
+    fn read_vec_small(&mut self, len: usize) -> Result<Vec<u8>> {
+        let mut result = Vec::with_capacity(len);
+        unsafe { result.set_len(len); }
+        try!(self.reader.read_exact(&mut result));
+        Ok(result)
+    }
+
+    fn read_vec_large(&mut self, mut len: usize, block: usize) -> Result<Vec<u8>> {
```

Comment:
Yes, will change it. Your version needs another cmp::min in
```rust
let mut result = Vec::with_capacity(BLOCK_SIZE);
```
or else even the smallest Vectors reserve space for at least `BLOCK_SIZE` bytes. I'll try to get around this.